src/x509.rs

```diff
@@ -76,13 +76,17 @@ impl<S: Integer, A: SignatureAlgorithm, K> TbsCertificate<S, A, K> {
     }
 }
 
-const TBS_CERTIFICATE_V3: u8 = 2;
+pub const TBS_CERTIFICATE_V1: u8 = 0;
+pub const TBS_CERTIFICATE_V2: u8 = 1;
+pub const TBS_CERTIFICATE_V3: u8 = 2;
 
 impl<S: DerWrite + Integer, A: DerWrite + SignatureAlgorithm, K: DerWrite> DerWrite
     for TbsCertificate<S, A, K> {
     fn write(&self, writer: DERWriter) {
         writer.write_sequence(|writer| {
-            writer.next().write_tagged(Tag::context(0), |w| TBS_CERTIFICATE_V3.write(w));
+            if self.version != TBS_CERTIFICATE_V1 { // default value
```

Comment:
I don't know enough about the X509 stuff going on here, but is it correct to not write the version number when it's the default?

Reply:
Yes, in this case it is. The TbsCertificate type is defined [here](https://www.rfc-editor.org/rfc/rfc5280#section-4.1) as
```
  TBSCertificate  ::=  SEQUENCE  {
        version         [0]  EXPLICIT Version DEFAULT v1,
        serialNumber         CertificateSerialNumber,
        signature            AlgorithmIdentifier,
        issuer               Name,
        validity             Validity,
        subject              Name,
        subjectPublicKeyInfo SubjectPublicKeyInfo,
        issuerUniqueID  [1]  IMPLICIT UniqueIdentifier OPTIONAL,
                             -- If present, version MUST be v2 or v3
        subjectUniqueID [2]  IMPLICIT UniqueIdentifier OPTIONAL,
                             -- If present, version MUST be v2 or v3
        extensions      [3]  EXPLICIT Extensions OPTIONAL
                             -- If present, version MUST be v3
  }
```
This means the `version` component has a tag of `0`, uses explicit tagging and has a default value of `v1`, which corresponds to the numerical value `0`.

Clause 11.5 of [X.690-0207](https://www.itu.int/ITU-T/studygroups/com17/languages/X.690-0207.pdf), which defines restrictions employed by CER and DER on top of BER, says:

> The encoding of a set value or sequence value shall not include an encoding for any component value which is equal to its default value.

Hence for DER encoding, we MUST not write the version component if it is equal to the default value.